### PR TITLE
support write data through UDP in sync mode

### PR DESCRIPTION
**Pull request by jiafu1115**
@majst01  

to avoid too much code change in one commit;
to avoid too many conflicts involved by other merged commits;
to make sure support UDP more comfortable and more safer;

Let's support UDP step by step:
(1) support sync or direct write string: this PR's focus
(2) support async write point or batch points

So you can keep the #226 opened for days after we completed two new PRs.

Now this PR only contain 88 line of source code change and 93 lines of test code added.  

Thanks for your review!

### CODE REVIEW

File: src/test/java/org/influxdb/InfluxDBTest.java [hunk 1]
```diff
@@ -151,6 +154,66 @@ public void testWriteStringData() {
     }
 
     /**
+     * Test writing to the database using string protocol through UDP.
+     */
+    @Test
+    public void testWriteStringDataThroughUDP() {
```
Comment (jiafu1115) [inline src/test/java/org/influxdb/InfluxDBTest.java]: there is some issue for test case through the udp function is right. I will figure out what
's the root cause.

Comment (jiafu1115) [inline src/test/java/org/influxdb/InfluxDBTest.java]: [query] 2016/11/09 01:24:10 CREATE DATABASE udp
[httpd] 10.140.202.187 - admin [09/Nov/2016:01:24:10 +0000] "POST /query?p=%5BREDACTED%5D&q=CREATE+DATABASE+%22udp%22&u=admin HTTP/1.1" 200 46 "-" "okhttp/3.4.2" 376d3267-a61b-11e6-8703-000000000000 23125
[query] 2016/11/09 01:24:10 SELECT * FROM udp.autogen.cpu GROUP BY *
[httpd] 10.140.202.187 - admin [09/Nov/2016:01:24:10 +0000] "GET /query?db=udp&p=%5BREDACTED%5D&q=SELECT+%2A+FROM+cpu+GROUP+BY+%2A&u=admin HTTP/1.1" 200 46 "-" "okhttp/3.4.2" 37784cb0-a61b-11e6-8705-000000000000 1319
[query] 2016/11/09 01:24:10 DROP DATABASE udp
[httpd] 10.140.202.187 - admin [09/Nov/2016:01:24:10 +0000] "POST /query?p=%5BREDACTED%5D&q=DROP+DATABASE+%22udp%22&u=admin HTTP/1.1" 200 46 "-" "okhttp/3.4.2" 377e834d-a61b-11e6-8706-000000000000 40067
[udp] 2016/11/09 01:24:11 failed to write point batch to database "udp": database not found: udp

**Comment by jiafu1115** [comment]
@majst01 WDYT?  we may need to handle PRs related with UDP ASAP due to it involved too much code change so that it easy conflict with other commits. 

Thanks in advanced.

**Comment by majst01** [comment]
I need to review, give me some time.

**Comment by jiafu1115** [comment]
@majst01 No problem. Just one reminder. Any issue please @ me. Thanks again.

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 2]
```diff
@@ -67,6 +77,14 @@ public InfluxDBImpl(final String url, final String username, final String passwo
     this.influxDBService = this.retrofit.create(InfluxDBService.class);
   }
 
+  private InetAddress parseHostAddress(final String url) {
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: This change is unrelated to udp support, as well as line 67

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 3]
```diff
@@ -212,6 +230,41 @@ public void write(final String database, final String retentionPolicy, final Con
...
+    byte[] bytes = records.getBytes(Charsets.UTF_8);
+    try {
+        datagramSocket.send(new DatagramPacket(bytes, bytes.length, hostAddress, udpPort));
+    } catch (IOException e) {
+        throw new RuntimeException(e);
+    }
+  }
+
+  private void initialDatagramSocket() {
+    if (datagramSocket == null)
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: missing Brackets

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 4]
```diff
@@ -212,6 +230,41 @@ public void write(final String database, final String retentionPolicy, final Con
...
+        synchronized (InfluxDBImpl.class) {
+            if (datagramSocket == null)
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: also missing brackets

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 5]
```diff
@@ -287,4 +340,18 @@ public void deleteDatabase(final String name) {
     }
   }
 
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void close() {
+    try {
+        this.disableBatch();
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: If we do so, Javadoc and Readme need to be adjusted as well.

File: src/main/java/org/influxdb/InfluxDB.java [hunk 6]
```diff
@@ -168,6 +168,22 @@ public void write(final String database, final String retentionPolicy,
                     final ConsistencyLevel consistency, final List<String> records);
 
   /**
+   * Write a set of Points to the influxdb database with the string records through UDP.
+   *
+   * @param udpPort
+   * @param records
+   */
+  public void write(final int udpPort, final String records);
```
Comment (majst01) [inline src/main/java/org/influxdb/InfluxDB.java]: Add some examples to README.md with these new methods as well please

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java (hunk 2 above)
Comment (jiafu1115) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: @majst01 No, We need it due to the followed method call:

datagramSocket.send(new DatagramPacket(bytes, bytes.length, **hostAddress**, udpPort));

I parse it in constructor to avoid parse again and again when send UDP package.

Thanks

File: src/main/java/org/influxdb/InfluxDB.java (hunk 6 above)
Comment (jiafu1115) [inline src/main/java/org/influxdb/InfluxDB.java]: @majst01  Can you change README.md after we complete both sync and async UDP support? It will be more better to express how to use it?  WDYT?

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java (hunk 2 above)
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: ok got it

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java [hunk 7]
```diff
@@ -212,6 +230,43 @@ public void write(final String database, final String retentionPolicy, final Con
    * {@inheritDoc}
    */
   @Override
+  public void write(final int udpPort, final String records) {
+    initialDatagramSocket();
+    byte[] bytes = records.getBytes(Charsets.UTF_8);
```
Comment (majst01) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: Javadoc need to mention that records have to be utf-8 encoded

Comment (jiafu1115) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: @majst01  Done with readme change and javadoc improve (for this method, it use {@inheritDoc} . the javadoc is located  in interface method)
 Thanks.

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java (hunk 5 above)
Comment (jiafu1115) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: @majst01 Done with readme change and javadoc improve (for this method, it use {@inheritDoc} . the javadoc is located in interface method)
Thanks for your review

File: src/main/java/org/influxdb/impl/InfluxDBImpl.java (hunk 7 above)
Comment (jiafu1115) [inline src/main/java/org/influxdb/impl/InfluxDBImpl.java]: @majst01  done, add into interface method.
